docs/adr/0012-runtime-message-results.md

```diff
@@ -0,0 +1,168 @@
+# ADR 0012: Runtime Message Results
+
+## Changelog
+
+- 2021-12-04: Initial version
+
+- 2021-12-10: Extend the implementation section
+
+## Status
+
+Proposed
+
+## Context
+
+Currently, the results of emitted runtime messages are `MessageEvent`s, which
+only provide information whether the message execution was successful or not.
+For various use-cases additional information about message results would be
+useful.
+
+One of such is supporting staking by runtimes. Currently, a runtime can emit an
+`AddEscrow` message, but is unaware of the actual amount of shares it obtained
+as a result of the added escrow. For some use-cases (e.g. runtime staking user
+deposited funds) this information is crucial for accounting.
+
+Similarly, for `ReclaimEscrow`, the runtime doesn't have the direct information
+at which epoch the stake gets debonded.
+
+The only way to currently obtain this data is to subscribe to consensus events,
+something which runtime doesn't have access to.
+
+Adding results to `MessageEvent` solves both of the mentioned use cases:
+
+- for `AddEscrow` the result should contain amount of shares obtained with the
+  escrow
+
+- for `ReclaimEscrow` the result should contain the amount of shares and epoch
+  at which the stake gets debonded
+
+## Decision
+
+Implement support for arbitrary result data in `MessageEvent` runtime message
+results.
+
+## Implementation
+
+- Result field is added to `roothash.MessageEvent` struct:
+
+```golang
+// MessageEvent is a runtime message processed event.
+type MessageEvent struct {
+        Module string `json:"module,omitempty"`
+        Code   uint32 `json:"code,omitempty"`
+        Index  uint32 `json:"index,omitempty"`
+
+        // Result contains message execution results for successfully executed messages.
+        Result cbor.RawMessage `json:"result,omitempty"
+}
+```
+
+The `Result` field is runtime message specific and is present only when the
+message execution was successful (`Code` is `errors.CodeNoError`).
+
+- `ExecuteMessage` method in `MessageSubscriber` interface is updated to include
+  a response:
+
+```golang
+// MessageSubscriber is a message subscriber interface.
+type MessageSubscriber interface {
+        // ExecuteMessage executes a given message.
+        ExecuteMessage(ctx *Context, kind, msg interface{}) (interface{}, error)
+}
+```
+
+- `Publish` method of the `MessageDispatcher` interface is updated to include
+  the response:
+
+```golang
+// MessageDispatcher is a message dispatcher interface.
+type MessageDispatcher interface {
+        // Publish publishes a message of a given kind by dispatching to all subscribers.
+        //
+        // In case there are no subscribers ErrNoSubscribers is returned.
+        Publish(ctx *Context, kind, msg interface{}) (interface{}, error)
```

Comment:
Should this instead be `[]interface{}` as there can be multiple subscribers? (Currently however there are not in this specific case used for runtime messages.)

And then the caller can decide what to do with the results (currently it would just take the first one).

Reply:
Yeah makes sense, updated.